controlplane/pkg/nsm/nsm_mechanims.go

```diff
@@ -4,88 +4,71 @@ import (
 				break
 			}
 		}
-		if c.Mechanism == nil {
-			return fmt.Errorf("Required mechanism are not found... %v ", r.MechanismPreferences)
-		}
-		if c.Mechanism.Parameters == nil {
-			c.Mechanism.Parameters = map[string]string{}
-		}
+	}
+
+	if connection.GetConnectionMechanism() == nil {
+		return fmt.Errorf("required mechanism are not found... %v ", request.GetRequestMechanismPreferences())
+	}
+
+	if connection.GetConnectionMechanism().GetParameters() == nil {
+		connection.GetConnectionMechanism().SetParameters(map[string]string{})
 	}
 
 	return nil
 }
 
-func (srv *networkServiceManager) selectRemoteMechanism(requestId string, request *remote_networkservice.NetworkServiceRequest, dp *model.Dataplane) (*remote_connection.Mechanism, error) {
-	for _, mechanism := range request.MechanismPreferences {
-		dp_mechanism := findRemoteMechanism(dp.RemoteMechanisms, remote_connection.MechanismType_VXLAN)
-		if dp_mechanism == nil {
+func (srv *networkServiceManager) selectRemoteMechanism(requestID string, request networkservice.Request, dp *model.Dataplane) (connection.Mechanism, error) {
+	for _, mechanism := range request.GetRequestMechanismPreferences() {
+		dpMechanism := findMechanism(dp.RemoteMechanisms, remote.MechanismType_VXLAN)
+		if dpMechanism == nil {
 			continue
 		}
+
 		// TODO: Add other mechanisms support
-		if mechanism.Type == remote_connection.MechanismType_VXLAN {
-			// Update DST IP to be ours
-			remoteSrc := mechanism.Parameters[remote_connection.VXLANSrcIP]
-			mechanism.Parameters[remote_connection.VXLANSrcIP] = remoteSrc
-			mechanism.Parameters[remote_connection.VXLANDstIP] = dp_mechanism.Parameters[remote_connection.VXLANSrcIP]
-			mechanism.Parameters[remote_connection.VXLANVNI] = strconv.FormatUint(uint64(srv.serviceRegistry.VniAllocator().Vni(dp_mechanism.Parameters[remote_connection.VXLANSrcIP], remoteSrc)), 10)
+
+		if mechanism.GetMechanismType() == remote.MechanismType_VXLAN {
+			parameters := mechanism.GetParameters()
+			dpParameters := dpMechanism.GetParameters()
+
+			parameters[remote.VXLANDstIP] = dpParameters[remote.VXLANSrcIP]
+
+			vni := srv.serviceRegistry.VniAllocator().Vni(dpParameters[remote.VXLANSrcIP], parameters[remote.VXLANSrcIP])
+			parameters[remote.VXLANVNI] = strconv.FormatUint(uint64(vni), 10)
 		}
-		logrus.Infof("NSM:(4.1-%v) Remote mechanism selected %v", requestId, mechanism)
+
+		logrus.Infof("NSM:(4.1-%v) Remote mechanism selected %v", requestID, mechanism)
 		return mechanism, nil
 	}
-	return nil, fmt.Errorf("NSM:(5.1-%v) Failed to select mechanism. No matched mechanisms found...", requestId)
-}
 
-func findRemoteMechanism(mechanismPreferences []*remote_connection.Mechanism, mechanismType remote_connection.MechanismType) *remote_connection.Mechanism {
-	for _, m := range mechanismPreferences {
-		if m.Type == mechanismType {
-			return m
-		}
-	}
-	return nil
+	return nil, fmt.Errorf("failed to select mechanism, no matched mechanisms found")
```

Comment:
`NSM:(5.1-%v) Failed to select mechanism. No matched mechanisms found...`

Reply:
Fixed on the upper level with:
```
return nil, fmt.Errorf("NSM:(5.1-%v) %v", requestID, err)
```